### PR TITLE
Added support for `DataType::Map` and `MapArray`

### PR DESCRIPTION
**Pull request by jorgecarleitao**
* physical and logical type
* FFI consume and produce
* IPC consume and produce
* integration tests unskipped

### CODE REVIEW

File: src/array/list/mod.rs [hunk 1]
```diff
@@ -129,19 +129,14 @@ impl<O: Offset> ListArray<O> {
     /// Returns the element at index `i`
     #[inline]
     pub fn value(&self, i: usize) -> Box<dyn Array> {
-        if self.is_null(i) {
```
Comment (nevi-me) [inline src/array/list/mod.rs:132]: Is this fine that you're no longer creating an empty list child?

File: src/array/map/mod.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    ) -> Self {
+        check_offsets(&offsets, field.len());
+
+        if let Some(ref validity) = validity {
+            assert_eq!(offsets.len() - 1, validity.len());
+        }
+
+        if let DataType::Struct(inner) = Self::get_field(&data_type).data_type() {
+            if inner.len() != 2 {
+                panic!("MapArray its inner `Struct` to have 2 fields (keys and maps)")
```
Comment (nevi-me) [inline src/array/map/mod.rs:74]:
```suggestion
                panic!("MapArray expects its inner `Struct` to have 2 fields (keys and maps)")
```

File: src/array/map/mod.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        let offset_1 = self.offsets[i + 1];
+        let length = (offset_1 - offset).to_usize();
+
+        // Safety:
+        // One of the invariants of the struct
+        // is that offsets are in bounds
+        unsafe { self.field.slice_unchecked(offset.to_usize(), length) }
+    }
+
+    /// Returns the element at index `i` as &str
```
Comment (nevi-me) [inline src/array/map/mod.rs:146]: `as &str`is this correct?

File: src/io/json_integration/schema.rs [hunk 4]
```diff
@@ -406,8 +407,16 @@ fn to_data_type(item: &Value, mut children: Vec<Field>) -> Result<DataType> {
             };
             DataType::Union(children, ids, is_sparse)
         }
+        "map" => {
+            let sorted_keys = if let Some(Value::Bool(sorted_keys)) = item.get("keysSorted") {
+                *sorted_keys
+            } else {
+                return Err(ArrowError::Schema("union requires mode".to_string()));
```
Comment (nevi-me) [inline src/io/json_integration/schema.rs:414]: nit: error not about keysSorted

File: src/scalar/mod.rs [hunk 5]
```diff
@@ -115,7 +115,7 @@ pub fn new_scalar(array: &dyn Array, index: usize) -> Box<dyn Scalar> {
         }
         FixedSizeBinary => todo!(),
         FixedSizeList => todo!(),
-        Union => todo!(),
+        Union | Map => todo!(),
```
Comment (nevi-me) [inline src/scalar/mod.rs:118]: I was working on map support, but didn't realise that I could skip a scalar as a start, so I got stuck here and didn't have a lot of time to continue.

I tried making a scalar to implement `HashMap` as a start, to validate that keys are hashable and unique.

**Review by nevi-me — APPROVED** [review]
Minor comments on some error messages

File: src/array/list/mod.rs (hunk 1 above)
Comment (jorgecarleitao) [reply src/array/list/mod.rs:132]: this was unrelated. The behavior of all others is that `value` returns the values irrespectively of the validity buffer. This keeps that behavior consistent.

File: src/array/mod.rs [hunk 6]
```diff
@@ -275,6 +276,7 @@ pub fn new_empty_array(data_type: DataType) -> Box<dyn Array> {
         FixedSizeList => Box::new(FixedSizeListArray::new_empty(data_type)),
         Struct => Box::new(StructArray::new_empty(data_type)),
         Union => Box::new(UnionArray::new_empty(data_type)),
+        Map => todo!(),
```
Comment (jorgecarleitao) [inline src/array/mod.rs:279]: this is wrong: we have the method

File: src/array/mod.rs [hunk 7]
```diff
@@ -304,6 +306,7 @@ pub fn new_null_array(data_type: DataType, length: usize) -> Box<dyn Array> {
         FixedSizeList => Box::new(FixedSizeListArray::new_null(data_type, length)),
         Struct => Box::new(StructArray::new_null(data_type, length)),
         Union => Box::new(UnionArray::new_null(data_type, length)),
+        Map => todo!(),
```
Comment (jorgecarleitao) [inline src/array/mod.rs:309]: this is wrong: we have the method

File: src/array/mod.rs [hunk 8]
```diff
@@ -341,6 +344,7 @@ pub fn clone(array: &dyn Array) -> Box<dyn Array> {
         FixedSizeList => clone_dyn!(array, FixedSizeListArray),
         Struct => clone_dyn!(array, StructArray),
         Union => clone_dyn!(array, UnionArray),
+        Map => todo!(),
```
Comment (jorgecarleitao) [inline src/array/mod.rs:347]: this is wrong: we have the method

File: src/compute/aggregate/memory.rs [hunk 9]
```diff
@@ -109,5 +109,6 @@ pub fn estimated_bytes_size(array: &dyn Array) -> usize {
         Dictionary(key_type) => with_match_physical_dictionary_key_type!(key_type, |$T| {
             dyn_dict!(array, $T)
         }),
+        Map => todo!(),
```
Comment (jorgecarleitao) [inline src/compute/aggregate/memory.rs:112]: this is wrong: we can compute this

File: src/scalar/mod.rs (hunk 5 above)
Comment (jorgecarleitao) [reply src/scalar/mod.rs:118]: yeah, the scalar API is always a bit behind: as long as we have the iterator API to iterate over individual elements, we are good to go.

**Comment by jorgecarleitao** [comment]
Thank you very much, @nevi-me for your review (great to see you :D)

**Pull request merged**